000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,447 @@
+
+### Listing offsets
+
+To list the current offsets the user will annotate the KafkaMirrorMaker2 resource with the new annotation `strimzi.io/connector-offsets` set to `list`, and the `strimzi.io/mirrormaker-connector` annotation.
+After the annotations are added, on the next reconciliation the operator will fetch the current offsets for the connector(s) and create Kubernetes ConfigMap(s) containing the JSON response(s).
+The operator will add an owner reference to the Kubernetes ConfigMap(s) pointing to the KafkaMirrorMaker2 resource that was annotated.
+If the ConfigMap(s) already exist the operator will patch the ConfigMap(s) with updated data and not add an owner reference.
+This means any existing keys in the ConfigMap(s) that the operator is not updating will remain as before.
+
+Since the operator reconciles connectors one-by-one, after each list operation the operator will update the `strimzi.io/mirrormaker-connector` annotation to remove the connector it has just listed offsets for.
+Once all list operations are complete the operator will then remove both the `strimzi.io/mirrormaker-connector` and `strimzi.io/connector-offsets` annotations from the KafkaMirrorMaker2 CR.
+This will make it easier to recover from errors.
+
+If the user wants to see an updated set of offsets they will need to re-annotate the resource.
+
+The name of the ConfigMap(s) the operator creates/updates will be set by the user in the KafkaMirrorMaker2 CR.
+The KafkaMirrorMaker2 CRD will be updated to contain a new connector property called `listOffsets` that must be set for the `strimzi.io/connector-offsets=list` annotation to take effect.
+The structure of the `listOffsets` property will be as below, to allow it to be extended in future if required:
+
+```yaml
+spec:
+  #...
+  mirrors:
+    - sourceConnector:
+        #...
+        listOffsets:
+          toConfigMap:
+            name: my-connector-offsets
+      checkpointConnector:
+        #...
+        listOffsets:
+          toConfigMap:
+            name: my-connector-offsets
+      heartbeatConnector:
+        #...
+        listOffsets:
+          toConfigMap:
+            name: my-connector-offsets
+```
+
+The ConfigMap(s) will contain a single entry for each connector, containing the response received from the GET /connectors/{connector}/offsets endpoint.
+This will make it possible to use the same ConfigMap for different connectors.
+The name of the entry will be based on the connector name, however since 
+[`>` characters are not allowed in ConfigMap keys](https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-object) the name will be altered to replace `->` with `,`.
+For example a connector called `east-kafka->west-kafka.MirrorSourceConnector` will have offsets placed under `east-kafka,west-kafka.MirrorSourceConnector.json`.
```

Comment:
Would it make more sense to just keep `-` for example? Using `,` as a separator seems strange in this place.

Reply:
The other option I thought of was changing the `->` to `--`, just because `-` often seems to be used in aliases, so I was trying to pick something that might make it easier to spot the separator

Reply:
I guess `--` would work for me. I can imagine `.` as well I guess as it is already used in the name. But `,` looks strange to me.